#it is a patch from linksys.It is needed for the light up the led.This patch was made for qsdk kernel (Qualcomm's downsteam kernel) 4.4.60 but it has been ported to 5.4(expect for the hackly code that for controling the color of led).
Index: linux-4.4/drivers/pinctrl/qcom/pinctrl-ipq5018.c
===================================================================
--- linux-4.4-old/drivers/pinctrl/qcom/pinctrl-ipq5018.c	2020-12-25 11:09:09.328382586 +0800
+++ linux-4.4/drivers/pinctrl/qcom/pinctrl-ipq5018.c	2020-12-25 11:09:56.097125434 +0800
@@ -613,7 +613,7 @@ static const char * const qdss_tracectl_
 	"gpio40",
 };
 static const char * const pwm0_groups[] = {
-	"gpio42",
+	"gpio42", "gpio46",
 };
 static const char * const qdss_cti_trig_out_b0_groups[] = {
 	"gpio42",
@@ -625,7 +625,7 @@ static const char * const xfem4_groups[]
 	"gpio42",
 };
 static const char * const pwm1_groups[] = {
-	"gpio43",
+	"gpio43", "gpio1",
 };
 static const char * const qdss_cti_trig_in_b0_groups[] = {
 	"gpio43",
@@ -649,7 +649,7 @@ static const char * const xfem6_groups[]
 	"gpio44",
 };
 static const char * const pwm3_groups[] = {
-	"gpio45",
+	"gpio45", "gpio30",
 };
 static const char * const qdss_cti_trig_in_b1_groups[] = {
 	"gpio45",
@@ -804,7 +804,7 @@ static const struct msm_function ipq5018
 static const struct msm_pingroup ipq5018_groups[] = {
 	PINGROUP(0, atest_char, _, qdss_cti_trig_out_a0, wci_txd, wci_rxd, xfem, _, _, _),
-	PINGROUP(1, atest_char, _, qdss_cti_trig_in_a0, wci_txd, wci_rxd, xfem, _, _, _),
+	PINGROUP(1, atest_char, pwm1, qdss_cti_trig_in_a0, wci_txd, wci_rxd, xfem, _, _, _),
	PINGROUP(2, atest_char, _, qdss_cti_trig_out_a1, wci_txd, wci_rxd, xfem, _, _, _),
	PINGROUP(3, atest_char, _, qdss_cti_trig_in_a1, wci_txd, wci_rxd, xfem, _, _, _),
	PINGROUP(4, sdc1_data, qspi_data, blsp1_spi1, btss, dbg_out, qdss_traceclk_a, _, bu      rn0, _),
	PINGROUP(5, sdc1_data, qspi_data, cxc_clk, blsp1_spi1, blsp1_i2c1, btss, _, qdss_tr      acectl_a, _),
@@ -860,7 +860,7 @@ static const struct msm_pingroup ipq5018
	PINGROUP(27, audio_txmclk, wsa_swrm, audio_txmclk, blsp2_spi, btss, _, qdss_traceda      ta_b, _, _),
	PINGROUP(28, audio_txbclk, wsa_swrm, blsp0_uart1, btss, qdss_tracedata_b, _, _, _,       _),
	PINGROUP(29, audio_txfsync, _, blsp0_uart1, _, qdss_tracedata_b, _, _, _, _),
-	PINGROUP(30, audio_txd, led2, led0, _, _, _, _, _, _),
+	PINGROUP(30, audio_txd, led2, led0, pwm3, _, _, _, _, _),
	PINGROUP(31, blsp2_spi0, blsp1_uart1, _, qdss_tracedata_b, eud_gpio, _, _, _, _),
	PINGROUP(32, blsp2_spi0, blsp1_uart1, _, qdss_tracedata_b, eud_gpio, _, _, _, _),
	PINGROUP(33, blsp2_i2c0, blsp2_spi0, blsp1_uart1, _, qdss_tracedata_b, eud_gpio, _,       _, _),
@@ -886,7 +886,7 @@ static const struct msm_pingroup ipq5018
	PINGROUP(43, pwm1, qdss_cti_trig_in_b0, wci_txd, wci_rxd, xfem, _, _, _, _),
	PINGROUP(44, pwm2, qdss_cti_trig_out_b1, wci_txd, wci_rxd, xfem, _, _, _, _),
	PINGROUP(45, pwm3, qdss_cti_trig_in_b1, wci_txd, wci_rxd, xfem, _, _, _, _),
-	PINGROUP(46, led0, _, _, _, _, _, _, _, _),
+	PINGROUP(46, led0, pwm0, _, _, _, _, _, _, _),
 };
